File: go/hamming/hamming.go
// Package hamming provides function to calculate the Hamming
// distance between two DNA strands
package hamming

import "errors"

// Distance receives two DNA strands and returns the Hamming distance
func Distance(a, b string) (distance int, err error) {
	if len(a) != len(b) {
		return 0, errors.New("DNA strands have different length")
	}
	if len(a) == 0 {
		return 0, nil
	}
	for index := 0; index < len(a); index++ {
		if a[index] != b[index] {
			distance++
		}
	}
	return distance, nil
}

File: ruby/hamming/hamming.rb
class Hamming
  def self.distance(strand1, strand2)
    if strand1.length != strand2.length
      raise ArgumentError, "Both strands don't have the same length"
    end

    strand1.length.times.count do |index|
      strand1[index] != strand2[index]
    end
  end
end
